### PR TITLE
prov/verbs: EP RDM generates CQ entries regarding to flags

### PR DESCRIPTION
**Pull request by evgeny-leksikov**
This series includes implementation of selective completion model and refactoring of tsend/trecv/rma nesting calls, i.e. now all simple functions call more complex ones. For example: tsend->tsendv->tsendmsg

@a-ilango , please review

Every commit passes build and fabtests

### CODE REVIEW

**Comment by lanl-ompi** [comment]
Test FAILed.

**Comment by j-xiong** [comment]
@evgeny-leksikov Could you put in the commit message some information about this patch? Also the changes look more than what the title says, could this be split into a few separate commits or PRs?

**Comment by evgeny-leksikov** [comment]
@j-xiong really, I squashed it from this [branch](https://github.com/evgeny-leksikov/libfabric/commits/completion_flags), but it looks too finely-granular (15 commits) and I'm not sure every intermediate commit is correct. If that variant is also not acceptable, I'll try reorganize the patch on 2 separate commits: refactoring and new functionality.

**Comment by j-xiong** [comment]
@evgeny-leksikov The 15 commits actually look better than the monolithic commit.  It would be even better if you could spend some time going over the commits, reorganizing when possible, correcting the typos, and try best to make each commit correct -- at least pass build, preferably pass fabtests.

**Comment by evgeny-leksikov** [comment]
@j-xiong Thanks! I'll update this PR tomorrow

**Comment by hppritcha** [comment]
ssh key problem at nersc
bot:retest

**Comment by lanl-ompi** [comment]
Test FAILed.

**Comment by hppritcha** [comment]
okay another try, somethings changed at nersc wrt using https protocol.
bot:retest

File: prov/verbs/src/verbs_rma.c [hunk 1]
```diff
@@ -508,7 +508,7 @@ static ssize_t fi_ibv_rdm_ep_rma_inject_write(struct fid_ep *ep,
 	struct fi_ibv_rdm_rma_start_data start_data = {
 		.conn = conn,
 		.ep_rdm = ep_rdm,
-		.data_len = (uint32_t)len,
+		.data_len = (uint64_t)len,
```
Comment (a-ilango) [inline prov/verbs/src/verbs_rma.c]: Why is this change required? I'm assuming this gets assigned to sge.length which is a uint32_t.

**Comment by j-xiong** [comment]
Please use a more descriptive line for the commit subject instead of a generic "refactoring".

**Comment by a-ilango** [comment]
Splitting the commits helped in understanding the individual changes but it would have been better (to read) if changes that are related to each other were merged. For example, lines added in one of the commits were removed in a later commit.

The changes in the PR mostly look okay to me.

**Comment by evgeny-leksikov** [comment]
@j-xiong do you mean make "one line comment"? Because a bit detailed description is in the second line of the comments but the main message is "there are nothing new", just a refactoring.

@a-ilango if it's not critical I would like to keep it as is for this pool request due to complexity, but I'll take it into account for the future

File: prov/verbs/src/verbs_rma.c (hunk 1 above)
Comment (evgeny-leksikov) [inline prov/verbs/src/verbs_rma.c]: In this case (inject) just to be aligned with fi_ibv_rdm_rma_start_data.data_len, msx_inject_size much less INT_MAX
But in common case, RDM/RMA currently supports more then 1GB transfers by segmentation.

**Comment by j-xiong** [comment]
@evgeny-leksikov The first line of the commit message is the "title". It's shown by commands like "git log --oneline" or "git shortlog". It should serve as a very brief summary of the commit. A single word "refactoring" is just too general. At least it can indicate what is being refactored.

**Comment by evgeny-leksikov** [comment]
@j-xiong i.e. to merge the comment into 1 line from 2 would be OK in this case?

**Comment by a-ilango** [comment]
> @a-ilango if it's not critical I would like to keep it as is for this pool request due to complexity, but I'll take it into account for the future

:+1:

**Comment by j-xiong** [comment]
@evgeny-leksikov that will be fine as long as the tile line doesn't become too long. The common practice is to keep the title line short, preferably less than 50 characters, but can be up to 75 characters if necessary.

**Comment by goodell** [comment]
The most common recommendations for git commit messages are documented [here](http://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html) and [here](http://chris.beams.io/posts/git-commit/), among many other places on the internet.

**Comment by evgeny-leksikov** [comment]
@j-xiong @goodell Thanks! Updated.

File: prov/verbs/src/ep_rdm/verbs_ep_rdm.c [hunk 2]
```diff
@@ -370,7 +371,7 @@ static int fi_ibv_rdm_ep_close(fid_t fid)
 	util_buf_pool_destroy(fi_ibv_rdm_tagged_request_pool);
 	util_buf_pool_destroy(fi_ibv_rdm_tagged_extra_buffers_pool);
 	util_buf_pool_destroy(fi_ibv_rdm_postponed_pool);
-
+
```
Comment (j-xiong) [inline prov/verbs/src/ep_rdm/verbs_ep_rdm.c]: Nit pick: extra tab/spaces in an empty line.

File: prov/verbs/src/verbs_rma.c [hunk 3]
```diff
@@ -276,22 +276,29 @@ fi_ibv_rdm_ep_rma_preinit(void **desc, void **raw_buf, size_t len,
...
+		uint64_t flags)
 {
-	struct fi_ibv_rdm_ep *ep = container_of(ep_fid, struct fi_ibv_rdm_ep,
-						ep_fid);
+	if(msg->iov_count != 1 || msg->rma_iov_count != 1) {
+		assert(0);
+		return -FI_EMSGSIZE;
+	}
+
+	struct fi_ibv_rdm_ep *ep =
```
Comment (j-xiong) [inline prov/verbs/src/verbs_rma.c]: mixed declaration and code is c99 feature.  are we going to force c99 compilation support when not really necessary?

Comment (goodell) [inline prov/verbs/src/verbs_rma.c]: While I agree mixed declarations/statements like this one should be converted to match the rest of the code base (and Linux kernel style guide), we definitely should not be worried about requiring common C99 compilation support.  For example, C99 designated initializers are an important feature that is widely supported and can greatly improve code safety.  I think that modern versions of MS Visual Studio even support them (>= VS2013? someone correct me if I'm wrong here).

**Comment by evgeny-leksikov** [comment]
Looks like something wrong with mac host - travis build was not completed

**Comment by j-xiong** [comment]
@a-ilango Do you have more comments before I merge this PR?

**Comment by a-ilango** [comment]
@j-xiong I don't have any more comments. Please go ahead.
